src/hotspot/share/utilities/lineReader.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "memory/allocation.inline.hpp"
+#include "runtime/os.hpp"
+#include "utilities/lineReader.hpp"
+
+LineReader::LineReader(const char* filename) : _filename(filename), _stream(nullptr) {
+  // Use os::open() because neither fopen() nor os::fopen()
+  // can handle long path name on Windows. HMM, is this still valid today???
+  int fd = os::open(filename, O_RDONLY, S_IREAD);
+  if (fd != -1) {
+    // Obtain a File* from the file descriptor so that getc()
+    // can be used in get_line().
+    _stream = os::fdopen(fd, "r");
+    if (_stream == nullptr) {
+      ::close(fd);
+    } else {
+      // fd will be closed by fclose(_stream)
+    }
+  } else {
+    _stream = nullptr;
+  }
+
+  _buffer_length = 32;
+  _buffer = NEW_RESOURCE_ARRAY(char, _buffer_length);
+}
+
+LineReader::~LineReader() {
+  close(); // Just in case
+}
+
+void LineReader::close() {
+  if (_stream != nullptr) {
+    fclose(_stream);
+    _stream = nullptr;
+  }
+}
+
+// Returns nullptr if we have reached EOF.
+// \n is treated as the line separator.
+// All occurrences of \r are stripped.
+char* LineReader::get_line() {
+  if (_stream == nullptr) {
+    return nullptr;
+  }
+  size_t buffer_pos = 0;
+  int c;
+  while ((c = getc(_stream)) != EOF) {
+    if (buffer_pos + 1 >= _buffer_length) {
+      size_t new_length = _buffer_length * 2;
```

Comment:
Again for a general utility this simple doubling of size may not be appropriate. For small size it should be okay though. And if we can set the initial size to avoid the need to grow then this is less of an issue.